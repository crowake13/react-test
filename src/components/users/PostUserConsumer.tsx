import React, { useContext, useEffect } from 'react';
import * as RENDER_LOG from '../../constants/render-log';
import { context } from '../../stores/users/post-user.context';
import { UsersContext } from '../../stores/users/users.context';
import withRenderLog from '../shared/withRenderLog';

const PostUserConsumer = ({ userId, loadingUserLabel, noUserLabel }: any) => {
  const postUserContext = useContext(context);
  const userService = useContext(UsersContext);

  useEffect(() => {
    if (userId && !postUserContext?.user) {
      userService.fetchById(userId);
    }
  }, [postUserContext, userId, userService]);

  return !postUserContext ? null : (
    <h6 className="card-subtitle mb-2 text-muted">
      {postUserContext.user?.name ??
        (postUserContext.isFetching
          ? loadingUserLabel
          : postUserContext.user?.name ?? noUserLabel)}
    </h6>
  );
};

export default withRenderLog({ greeting: RENDER_LOG.GREETING })(
  PostUserConsumer
);
