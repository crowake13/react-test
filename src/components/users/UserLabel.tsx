import React from 'react';
import * as RENDER_LOG from '../../constants/render-log';
import withRenderLog from '../shared/withRenderLog';

export const UserLabel = ({ label }: { label: string }) => {
  return <h6 className="card-subtitle mb-2 text-muted">{label}</h6>;
};

export default withRenderLog({ greeting: RENDER_LOG.GREETING })(UserLabel);
