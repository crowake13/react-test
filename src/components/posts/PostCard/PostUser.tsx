import React, { useContext, useEffect } from 'react';
import * as RENDER_LOG from '../../../constants/render-log';
import { ID } from '../../../stores/entities/entity.facade';
import PostUserContext from '../../../stores/users/post-user.context';
import { UsersContext } from '../../../stores/users/users.context';
import withRenderLog from '../../shared/withRenderLog';

interface PostUserProps {
  userId: ID;
  loadingUserLabel: string;
  noUserLabel: string;
}

const PostUser = ({ userId, loadingUserLabel, noUserLabel }: PostUserProps) => {
  const { user, isFetching } = useContext(PostUserContext);
  const userService = useContext(UsersContext);

  useEffect(() => {
    if (userId && !user) {
      userService.fetchById(userId);
    }
  }, [user, userId, userService]);

  const label = !user
    ? isFetching
      ? loadingUserLabel
      : noUserLabel
    : user.name;

  return (
    <h6 title={label} className="card-subtitle mb-2 text-muted line-clamp-1">
      {label}
    </h6>
  );
};

export default withRenderLog({ greeting: RENDER_LOG.GREETING })(PostUser);
