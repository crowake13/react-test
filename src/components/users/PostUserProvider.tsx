import { useObservable } from '@mindspace-io/utils';
import React, { useContext } from 'react';
import * as RENDER_LOG from '../../constants/render-log';
import { useFetch } from '../../hooks/fetch.hook';
import { ID } from '../../stores/entities/entity.facade';
import PostUserContext from '../../stores/users/post-user.context';
import { UsersContext } from '../../stores/users/users.context';
import withRenderLog from '../shared/withRenderLog';

const PostUserProvider = ({
  id,
  slug,
  children
}: {
  id: ID;
  slug: string;
  children?: any;
}) => {
  const usersService = useContext(UsersContext);
  const [isFetching] = useFetch(usersService, slug);
  const [users] = useObservable(usersService.entitie$, []);

  const user = users.filter(entity => entity.id === id)[0] ?? null;

  return (
    <PostUserContext.Provider
      value={{
        user,
        isFetching
      }}
    >
      {children}
    </PostUserContext.Provider>
  );
};

export default withRenderLog({ greeting: RENDER_LOG.GREETING })(
  PostUserProvider
);
