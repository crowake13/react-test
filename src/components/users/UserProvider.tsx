import React, { useContext } from 'react';
import * as RENDER_LOG from '../../constants/render-log';
import { useFetch } from '../../hooks/fetch.hook';
import { ID } from '../../stores/entities/entity.facade';
import { Provider } from '../../stores/users/post-user.context';
import { UsersContext } from '../../stores/users/users.context';
import withRenderLog from '../shared/withRenderLog';

const UserProvider = ({
  id,
  slug,
  children
}: {
  id: ID;
  slug: string;
  children?: any;
}) => {
  const usersService = useContext(UsersContext);
  const [isFetching] = useFetch(usersService, slug);

  return (
    <Provider
      value={{
        user: usersService.getById(id),
        isFetching: isFetching
      }}
    >
      {children}
    </Provider>
  );
};

export default withRenderLog({ greeting: RENDER_LOG.GREETING })(UserProvider);
