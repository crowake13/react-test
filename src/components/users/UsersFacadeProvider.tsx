import React from 'react';
import * as RENDER_LOG from '../../constants/render-log';
import { UsersContext } from '../../stores/users/users.context';
import { facade, IUsersFacade } from '../../stores/users/users.facade';
import withRenderLog from '../shared/withRenderLog';

interface UsersFacadeProviderProps extends Partial<IUsersFacade> {
  children: any;
}

const UsersFacadeProvider = ({
  children,
  ...props
}: UsersFacadeProviderProps) => {
  return (
    <UsersContext.Provider
      value={{
        searchTerm: props.searchTerm ?? facade.searchTerm,
        searchTerm$: props.searchTerm$ ?? facade.searchTerm$,
        getFilteredUsers: props.getFilteredUsers ?? facade.getFilteredUsers,
        updateSearchTerm: props.updateSearchTerm ?? facade.updateSearchTerm,
        errors: props.errors ?? facade.errors,
        error$: props.error$ ?? facade.error$,
        active$: props.active$ ?? facade.active$,
        activate: props.activate ?? facade.activate,
        getAll: props.getAll ?? facade.getAll,
        getById: props.getById ?? facade.getById,
        entitie$: props.entitie$ ?? facade.entitie$,
        isFetchingBySlug: props.isFetchingBySlug ?? facade.isFetchingBySlug,
        isFetching: props.isFetching ?? facade.isFetching,
        isFetching$: props.isFetching$ ?? facade.isFetching$,
        fetch: props.fetch ?? facade.fetch,
        fetchAll: props.fetchAll ?? facade.fetchAll,
        fetchById: props.fetchById ?? facade.fetchById
      }}
    >
      {children}
    </UsersContext.Provider>
  );
};

export default withRenderLog({ greeting: RENDER_LOG.GREETING })(
  UsersFacadeProvider
);
