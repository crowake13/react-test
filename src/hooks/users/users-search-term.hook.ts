import { useObservable } from '@mindspace-io/utils';
import { useContext } from 'react';
import { UsersContext } from '../../stores/users/users.context';

export type UsersSearchTermHookDuple = [string, (text: string) => void];

/**
 * Custom Hook to manage a view Model for User view components
 */
export function useUsersSearchTerm(): UsersSearchTermHookDuple {
  const usersService = useContext(UsersContext);

  const [searchTerm] = useObservable(
    usersService.searchTerm$,
    usersService.searchTerm
  );

  return [searchTerm, usersService.updateSearchTerm.bind(usersService)];
}
