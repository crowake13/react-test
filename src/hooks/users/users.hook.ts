import { useObservable } from '@mindspace-io/utils';
import { useContext } from 'react';
import { User } from '../../stores/users/user.model';
import { UsersContext } from '../../stores/users/users.context';
import { IUsersFacade } from '../../stores/users/users.facade';

export type UsersHookTuple = [User[], IUsersFacade];

/**
 * Custom Hook to manage a view Model for User view components
 */
export function useUsers(): UsersHookTuple {
  const usersService = useContext(UsersContext);

  const [users] = useObservable(usersService.entitie$, []);

  return [users, usersService];
}
