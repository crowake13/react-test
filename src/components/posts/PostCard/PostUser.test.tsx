import React from 'react';
import { render, unmountComponentAtNode } from 'react-dom';
import { act } from 'react-dom/test-utils';
import { Subject } from 'rxjs';
import { filter, map } from 'rxjs/operators';
import { ID } from '../../../stores/entities/entity.facade';
import { User } from '../../../stores/users/user.model';
import { facade } from '../../../stores/users/users.facade';
import UserProvider from '../../users/UserProvider';
import UsersFacadeProvider from '../../users/UsersFacadeProvider';
import PostUser from './PostUser';

let container: HTMLElement | null = null;

beforeEach(() => {
  // setup a DOM element as a render target
  container = document.createElement('div');
  document.body.appendChild(container);
});

afterEach(() => {
  if (!container) {
    return;
  }

  // cleanup on exiting
  unmountComponentAtNode(container);
  container.remove();
  container = null;
});

const users: User[] = [
  {
    id: 1,
    name: 'Leanne Graham',
    username: 'Bret',
    email: 'Sincere@april.biz',
    address: {
      street: 'Kulas Light',
      suite: 'Apt. 556',
      city: 'Gwenborough',
      zipcode: '92998-3874',
      geo: {
        lat: -37.3159,
        lng: 81.1496
      }
    },
    phone: '1-770-736-8031 x56442',
    website: 'hildegard.org',
    company: {
      name: 'Romaguera-Crona',
      catchPhrase: 'Multi-layered client-server neural-net',
      bs: 'harness real-time e-markets'
    }
  }
];

it('renders PostUser manually', () => {
  const _entitie$ = new Subject<User[]>();
  const entitie$ = _entitie$.asObservable();

  const _isFetching$ = new Subject<{ [key: string]: boolean }>();
  const isFetching$ = _isFetching$.asObservable();

  const setIsFetching = (value: boolean, slug: string = 'users') => {
    _isFetching$.next({
      [slug]: value
    });
  };

  const getById = (id: ID) => {
    return users[0].id === id ? users[0] : null;
  };

  act(() => {
    render(
      <UsersFacadeProvider
        entitie$={entitie$}
        isFetching$={isFetching$}
        getById={getById}
      >
        <UserProvider id={users[0].id} slug="users">
          <PostUser
            loadingUserLabel="Author is loading..."
            noUserLabel="Author could not be found!"
          />
        </UserProvider>
      </UsersFacadeProvider>,
      container
    );
    setIsFetching(false);
  });
  expect(container?.innerHTML).toMatchSnapshot();

  act(() => {
    setIsFetching(true);
  });

  expect(container?.innerHTML).toMatchSnapshot();

  act(() => {
    _entitie$.next(users);
    setIsFetching(false);
  });

  expect(container?.innerHTML).toMatchSnapshot();
});

it('renders PostUser with fetch', async () => {
  jest.spyOn(global, 'fetch' as any).mockImplementation(() =>
    Promise.resolve({
      json: () => {
        return Promise.resolve(users);
      }
    })
  );

  // Use the asynchronous version of act to apply resolved promises
  act(() => {
    render(
      <UsersFacadeProvider {...facade}>
        <UserProvider id={users[0].id} slug="users">
          <PostUser
            loadingUserLabel="Author is loading..."
            noUserLabel="Author could not be found!"
          />
        </UserProvider>
      </UsersFacadeProvider>,
      container
    );
  });
  expect(container?.innerHTML).toMatchSnapshot();

  act(() => {
    facade.fetch(`users`);
  });

  await act(async () => {
    facade.isFetching$
      .pipe(
        map(isFetching => isFetching[`users`]),
        filter(isFetchingUsers => isFetchingUsers)
      )
      .toPromise();
  });

  expect(container?.innerHTML).toMatchSnapshot();

  await act(async () => {
    facade.isFetching$
      .pipe(
        map(isFetching => isFetching[`users`]),
        filter(isFetchingUsers => !isFetchingUsers)
      )
      .toPromise();
  });

  expect(container?.innerHTML).toMatchSnapshot();

  // remove the mock to ensure tests are completely isolated
  (global as any).fetch.mockRestore();
});
