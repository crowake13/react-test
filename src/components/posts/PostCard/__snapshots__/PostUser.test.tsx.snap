// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders PostUser manually 1`] = `"<h6 title=\\"Author could not be found!\\" class=\\"card-subtitle mb-2 text-muted line-clamp-1\\">Author could not be found!</h6>"`;

exports[`renders PostUser manually 2`] = `"<h6 title=\\"Author is loading...\\" class=\\"card-subtitle mb-2 text-muted line-clamp-1\\">Author is loading...</h6>"`;

exports[`renders PostUser manually 3`] = `"<h6 title=\\"Leanne Graham\\" class=\\"card-subtitle mb-2 text-muted line-clamp-1\\">Leanne Graham</h6>"`;
